compiler/circle-mpqsolver/src/bisection/BisectionSolver.cpp

```diff
@@ -0,0 +1,206 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "BisectionSolver.h"
+#include "DepthParameterizer.h"
+#include "ErrorMetric.h"
+
+#include <luci/ImporterEx.h>
+#include <luci/Log.h>
+
+#include <cmath>
+#include <iostream>
+
+using namespace mpqsolver::bisection;
+
+namespace
+{
+
+std::unique_ptr<luci::Module> read_module(const std::string &path)
+{
+  luci::ImporterEx importerex;
+  auto module = importerex.importVerifyModule(path);
+  if (module.get() == nullptr)
+  {
+    std::cerr << "ERROR: Failed to load " << path << std::endl;
+    return nullptr;
+  }
+
+  return module;
+}
+
+} // namespace
+
+BisectionSolver::BisectionSolver(const std::string &input_data_path, float qerror_ratio,
+                                 const std::string &input_quantization,
+                                 const std::string &output_quantization)
+  : MPQSolver(input_data_path, qerror_ratio, input_quantization, output_quantization)
+{
+  _quantizer = std::make_unique<Quantizer>(_input_quantization, _output_quantization);
+}
+
+float BisectionSolver::evaluate(const DatasetEvaluator &evaluator, const std::string &flt_path,
+                                const std::string &def_quant, LayerParams &layers)
+{
+  auto model = read_module(flt_path);
+  // get fake quantized model for evaluation
+  if (!_quantizer->fake_quantize(model.get(), def_quant, layers))
+  {
+    throw std::runtime_error("Failed to produce fake-quantized model.");
+  }
+
+  return evaluator.evaluate(model.get());
+}
+
+void BisectionSolver::algorithm(Algorithm algorithm) { _algorithm = algorithm; }
+
+std::unique_ptr<luci::Module> BisectionSolver::run(const std::string &module_path)
+{
+  LOGGER(l);
+
+  auto module = read_module(module_path);
+
+  float min_depth = 0.f;
+  float max_depth = 0.f;
+  NodeDepthType nodes_depth;
+  if (compute_depth(module.get(), nodes_depth, min_depth, max_depth) !=
+      ParameterizerResult::SUCCESS)
+  {
+    std::cerr << "ERROR: Invalid graph for bisectioning" << std::endl;
+    return nullptr;
+  }
+
+  std::unique_ptr<MAEMetric> metric = std::make_unique<MAEMetric>();
+  DatasetEvaluator evaluator(module.get(), _input_data_path, *metric.get());
+
+  LayerParams layer_params;
+  float int16_qerror =
+    evaluate(evaluator, module_path, "int16" /* default quant_dtype */, layer_params);
+  VERBOSE(l, 0) << "Full int16 model quantization error " << int16_qerror << std::endl;
+
+  float uint8_qerror =
+    evaluate(evaluator, module_path, "uint8" /* default quant_dtype */, layer_params);
+  VERBOSE(l, 0) << "Full uint8 model quantization error " << uint8_qerror << std::endl;
+
+  if (int16_qerror > uint8_qerror)
+  {
+    throw std::runtime_error("Q8 model's qerror is less than Q16 model's qerror.");
```

Comment:
Q) throw with `std::runtime_error` and showing message with `std::cerr` + `return nullptr` are mix used. is there any particular reason doing so?

Reply:
`throw` is used for extremely rare cases, while `std::cerr` are used for comfortable use, so that we can see the reason of error immediately.